### ISSUE TITLE
backup/restore db

### ISSUE THREAD

**Issue by chengweiv5** [sameersbn/docker-gitlab#456]
Since I didn't find a good place to ask questions, so here I came.

According to backup/restore sections in README.md, backup operation will backup all includes db and I also confirmed that from a backup tar, there is a db/ directory and a .sql file in it.

I'm using gitlab 7.9.2 and linked postgresql & redis with their own db volumes.

Now I'm trying to upgrade to gitlab 8.0.4 and wondering if I does restore from backup created by rake task, what will happen? does gitlab will re-write all db from the backup? or just use db in postgresql & redis?

**Comment by chengweiv5** [comment]
I put a backup tarball into gitlab 8.0.4 backup directory and run `restore` task like below, but it fails.

```
# docker exec -it gitlab
# cd /home/git/gitlab

root@c649fcea323c:/home/git/gitlab# /sbin/entrypoint.sh app:rake gitlab:backup:restore BACKUP=1444248015   
Running "gitlab:backup:restore" rake task ...
Unpacking backup ... done
GitLab version mismatch:
  Your current GitLab version (8.0.4) differs from the GitLab version in the backup!
  Please switch to the following version and try again:
  version: 7.9.2

Hint: git checkout v7.9.2
```

So how could I upgrade gitlab from v7.9.2 to the latest 8.0.4?

**Comment by chengweiv5** [comment]
And I found that gitlab documented that https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/raketasks/backup_restore.md
> Restore a previously created backup
You can only restore a backup to exactly the same version of GitLab that you created it on, for example 7.2.1.

And in our document https://github.com/sameersbn/docker-gitlab#upgrading, it doesn't say anything about restore data from backup after new version of gitlab started. Seems it does a in-place upgrade? just reuse volumes mounted by previous gitlab container? Does this means all data are compatible between different gitlab versions?

**Comment by chengweiv5** [comment]
OK, I found document about how to upgrade from source-installation

https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/update/patch_versions.md
https://gitlab.com/gitlab-org/gitlab-ce/tree/master/doc/update

**Comment by sameersbn** [comment]
@chengweiv5 that is true a backup can only be restored on the same gitlab version. You can upgrade to Gitlab `8.0` only from `v7.14.3`. So you fist need to upgrade to `v7.14.3` and then upgrade to `v8.0.4`.

**Comment by chengweiv5** [comment]
@sameersbn is there any inside about why I have to first upgrade to v7.14.3? I didn't find any from gitlab document above, thanks in advance!

**Comment by sameersbn** [comment]
the CI migration guide has such a requirement http://doc.gitlab.com/ce/migrate_ci_to_ce/README.html
I just assume thats the case for GitLab. If you are facing issues migrating to 8.0 from 7.9.2, then I would recommend upgrading to 7.14.3 first.

**Comment by favorinfo** [comment]
I've got the same problem. 
As I upgrade the gitlab to the newest version, I found that I can visit the site properly. However once I want to visit the repos, it came out the `repository: does not exist` .
I search the repos in my disk, and then found that there is a directory named '`gitlab-satellites`' in '`/home/git/data/`'. However the default GITLAB_REPOS_DIR in 8.0.4 is `/home/git/data/repository`. is that matters?

**Comment by chengweiv5** [comment]
@sameersbn thanks, I'll check that document, BTW, I didn't use gitlab-ci so far.

**Comment by sameersbn** [comment]
@favorinfo do you see a `repositories` directory in `/home/git/data`? If yes, inspect the contents to see ensure that your repos exist.

Now, stop your gitlab container and start again adding `app:sanitize` at the end of the docker run command. It will ensure that all files and directories have the right permissions. Once completed, start gitlab normally and check if the issue is resolved.

**Comment by favorinfo** [comment]
thx . I found that there is a `repositories` in `/home/git/data`.  the content in it  just like the '.git' dir in git repos. and files in `gitlab-satellites` are the repo files.  what if I copy  the files in  `gitlab-satellites` to the dir `repository`?

**Comment by sameersbn** [comment]
yea.. they will be like `/home/git/data/repositories/docker/gitlab.git`. If you drop into the directories you should be able to do a `git log`, `git tag`, etc. If these commands work, means your repos are there and probably its only a permissions issue. Nothing that `app:sanitize` can't resolve.

**Comment by sameersbn** [comment]
@favorinfo you could try that.

**Comment by sameersbn** [comment]
@favorinfo after the rollback, please check if you can browse  the repos

**Comment by favorinfo** [comment]
I copy all the files in `gitlab-satellites` to `repository` and  use `/sbin/entrypoint.sh app:sanitiz` , then solved all the problems. 
thx!

**Comment by sameersbn** [comment]
I don't think that is the proper solution. The repos in the 
`repositories` directory are bare repos. The ones in `gitlab-satellites` 
are not.

In the worst case, I would use the repos from `gitlab-satellite` and 
push them to the master using `git push` instead of doing what you just 
did :)

On Friday 09 October 2015 01:33 PM, favorinfo wrote:
> I copy all the files in |gitlab-satellites| to |repository| and use
> |/sbin/entrypoint.sh app:sanitiz| , then solved all the problems.
> thx!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sameersbn/docker-gitlab/issues/456#issuecomment-146790931>.
>

-- 
Sameer Naik

w: http://www.damagehead.com e: sameer@damagehead.com

**Comment by chengweiv5** [comment]
@favorinfo I did below to upgrade 7.9.2 to 8.0.4, **note: I didn't use gitlab-ci before**

Before upgrade, there are three containers started by docker-compose
- gitlab
- postgresql
- redis

To **verify** if upgrade works, I backuped all data in gitlab persistent store(includes postgresql volume and redis volume) and untar in another machine, where I will start docker-compose to create the 3 containers.

So now I have all the data in another machine, and then I start docker-compose, after all of them started, I opened browser to access gitlab and all works great!

To ensure all goes fine, I did below commands as [gitlab document suggested](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/update/patch_versions.md), of course in the gitlab container (`docker exec -it gitlab /bin/bash`)

```
sudo -u git -H bundle exec rake gitlab:env:info RAILS_ENV=production
sudo -u git -H bundle exec rake gitlab:check RAILS_ENV=production
```

there is only a small problem says that uploads isn't in 0755 mode, so I just did chmod to fix it.

**BTW**, I didn't do any migration commands as [gitlab document suggested](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/update/patch_versions.md) in step 4.

**Comment by chengweiv5** [comment]
OK, entrypoint.sh does step 4 for me, great! So it works out of box.

**Comment by chengweiv5** [comment]
FYI, I upgraded to 8.0.4 directly minutes ago, all goes fine.

**Comment by sameersbn** [comment]
`app:sanitize` will now (properly) set the ownership and permissions of the `uploads/` directory

**Comment by eanbowman** [comment]
> 
> 
> @chengweiv5 that is true a backup can only be restored on the same gitlab version. You can upgrade to Gitlab `8.0` only from `v7.14.3`. So you first need to upgrade your existing setup to `v7.14.3` and then upgrade to `v8.0.4`. There is no requirement to restore a backup while upgrading. The backup is only a fail-safe so that you can rollback if things go south.

What fucking good is a backup then?

I have the old drive my gitlab server was on. Still working. Want to restore the repositories from it to my new drive.

How do I do THAT?
